### PR TITLE
Support for renaming constants

### PR DESCRIPTION
**Pull request by vilhena-stripe**
This is largely based off @jvilk-stripe's work in https://github.com/sorbet/sorbet/compare/jvilk-rename-constant

### Motivation
Renaming Ruby constants can be a pain in `pay-server`, but Sorbet already has all the information to do the rename. We want to leverage this information and integrate renaming into VS Code

### Test plan
See included automated tests.

### CODE REVIEW

File: test/helpers/position_assertions.cc [hunk 1]
```diff
@@ -1233,6 +1239,112 @@ string ApplyCompletionAssertion::toString() const {
...
+        // Get the file path from the edited document's uri so we can determine the path to the .rbedited file
+        index = renameItem->textDocument->uri.find(testDataPath, 0);
+        sourceFilePath = renameItem->textDocument->uri.substr(index, renameItem->textDocument->uri.length() - 1);
+        expectedEditedFilePath = updatedFilePath(sourceFilePath, this->version);
+
+        auto &edits = renameItem->edits;
+        try {
+            expectedEditedFileContents = FileOps::read(expectedEditedFilePath);
+        } catch (FileNotFoundException e) {
+            ADD_FAIL_CHECK_AT(filename.c_str(), this->assertionLine + 1,
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1316]: I think you also need to check that all files which have a `this->version` version have been edited.

e.g., if you have `test__foo.rb`, `test__foo.A.rbedited`, `test__bar.rb`, and `test__bar.A.rbedited`, and rename only renames `foo`, the test should fail.

File: test/helpers/position_assertions.cc [hunk 2]
```diff
@@ -1233,6 +1239,112 @@ string ApplyCompletionAssertion::toString() const {
...
+
+        actualEditedFileContents = string(sourceFiles[sourceFilePath]);
+        for (auto &edit : edits) {
+            auto file = core::File(string(sourceFilePath), string(actualEditedFileContents), core::File::Type::Normal);
+
+            actualEditedFileContents = applyEdit(actualEditedFileContents, file, *edit->range, edit->newText);
+        }
+
+        {
+            INFO(fmt::format("The expected (rbedited) file contents for this rename did not match the actual file "
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1336]: Can you use `ADD_FAIL_CHECK_AT` and print the filename, expected file contents, and the actual file contents?

e.g.:

https://github.com/sorbet/sorbet/blob/b21d4db9/test/helpers/position_assertions.cc#L1138

File: test/helpers/position_assertions.cc [hunk 3]
```diff
@@ -1233,6 +1239,112 @@ string ApplyCompletionAssertion::toString() const {
...
+                             "contents post-edit of {} ",
+                             expectedEditedFilePath));
+            CHECK_EQ(expectedEditedFileContents, actualEditedFileContents);
+        }
+    }
+}
+
+string ApplyRenameAssertion::toString() const {
+    return fmt::format("apply-rename: [{}] item: {}", version, index);
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1345]:
```suggestion
    return fmt::format("apply-rename: [{}] newName: {}", version, newName);
```

Comment (jvilk-stripe) [reply test/helpers/position_assertions.cc:1345]: Also... how is this compiling; what is `index`?

**Review by jvilk-stripe — CHANGES_REQUESTED** [review]
Woo!!! This is looking great.

I left some comments, mostly pertaining to edge cases. Hopefully they make sense!

For tests, I suspect we'll want to test errors/invalid rename locations/etc, which may require adding an extra parameter to the rename test assertion.

File: main/lsp/requests/rename.cc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+unique_ptr<WorkspaceEdit> RenameTask::getRenameEdits(LSPTypecheckerDelegate &typechecker, core::SymbolRef symbol,
+                                                     std::string_view newName) {
+    const core::GlobalState &gs = typechecker.state();
+    vector<unique_ptr<Location>> references;
+    references = getReferencesToSymbol(typechecker, symbol);
+    auto originalName = symbol.data(gs)->name.toString(gs);
+    auto we = make_unique<WorkspaceEdit>();
+    // TextEdit
```
[the remark this answers is not in the document]
Comment (vilhena-stripe) [reply main/lsp/requests/rename.cc:32]: oops, this was a note to myself

File: test/helpers/position_assertions.cc [hunk 5]
```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
...
+        return;
+    }
+
+    {
+        INFO("doTextDocumentRename failed; see error above.");
+        REQUIRE_NE(workspaceEdits, nullptr);
+    }
+
+    auto &renameItems = *workspaceEdits->documentChanges;
+    auto it = sourceFileContents.find(this->filename);
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1295]: Minor nit: Is this variable used anywhere? I think you can replace this with an `if (sourceFileContents.find(this->filename) == sourceFileContents.end() {`

File: test/helpers/position_assertions.cc [hunk 6]
```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
...
+    auto testDataPath = this->filename.substr(0, index);
+    // map of all .rb and .rbi files
+    UnorderedMap<string, string> sourceFiles;
+    // map of all .rbedited files whose version equals `this->version`
+    UnorderedMap<string, string> expectedEditedFiles;
+    for (auto filePath : FileOps::listFilesInDir(testDataPath, {".rb", ".rbi", ".rbedited"}, false, {}, {})) {
+        auto extension = FileOps::getExtension(filePath);
+        if (extension == "rbedited") {
+            auto extensionIndex = filePath.rfind(".rbedited", filePath.length());
+            auto fileVersion = string(FileOps::getExtension(filePath.substr(0, extensionIndex)));
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1311]: Ah, so this caused an issue before because you are:

1. Calling `filePath.substr`, which creates a new `string` (not a `string_view` of `filePath`
2. `FileOps::getExtension` is returning a `string_view` of that new string
3. The new string is destructed after this line since it's not stored anywhere.

The new call to `string` seems to properly copy the string before the destruction.

File: test/helpers/position_assertions.cc [hunk 7]
```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
...
+            if (fileVersion == this->version) {
+                expectedEditedFiles[filePath] = FileOps::read(filePath);
+            }
+        } else {
+            sourceFiles[filePath] = FileOps::read(filePath);
+        }
+    }
+
+    string sourceFilePath;
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1320]: Minor nit: These seem to only be used in the `for` loop; is there any reason why they are defined out here rather than at their first definition?

File: test/helpers/position_assertions.cc [hunk 8]
```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
...
+        }
+        sourceFiles[sourceFilePath] = actualEditedFileContents;
+    }
+
+    // Compare every source file to its .rbedited of the same version if one exists
+    // There are 3 cases we're handling
+    // 1. present in bothsource and edited maps and contents match => success
+    // 2. present in original content, but not edited => fails
+    // 3. unexpected edit => fails
+    for (auto &it : sourceFiles) {
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1360]: You can actually do the following!

```suggestion
    for (auto &[filePath, actualEditedFileContents] : sourceFiles) {
```

File: test/helpers/position_assertions.cc [hunk 9]
```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
...
+        string filePath = it.first;
+        actualEditedFileContents = it.second;
+        expectedEditedFilePath = updatedFilePath(filePath, this->version);
+        expectedEditedFileContents = expectedEditedFiles[expectedEditedFilePath];
+        if (expectedEditedFileContents.empty()) {
+            continue;
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1366]: Isn't this ignoring case 3 (unexpected edit)?

File: test/helpers/position_assertions.cc (hunk 2 above)
Comment (jvilk-stripe) [reply test/helpers/position_assertions.cc:1369]: We decided we will spruce up error messages in a subsequent PR.

File: test/helpers/position_assertions.cc [hunk 10]
```diff
@@ -1218,6 +1223,162 @@ string ApplyCompletionAssertion::toString() const {
...
+        for (auto &edit : edits) {
+            auto file = core::File(string(sourceFilePath), string(actualEditedFileContents), core::File::Type::Normal);
+
+            actualEditedFileContents = applyEdit(actualEditedFileContents, file, *edit->range, edit->newText);
+        }
+        sourceFiles[sourceFilePath] = actualEditedFileContents;
+    }
+
+    // Compare every source file to its .rbedited of the same version if one exists
+    // There are 3 cases we're handling
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.cc:1356]: I guess there is four, with four being present in both source and edited maps and contents do not match.

File: test/helpers/position_assertions.cc (hunk 9 above)
Comment (jvilk-stripe) [reply test/helpers/position_assertions.cc:1366]: You are also not handling case 2 (something is in expectedEdited but not in sourceFiles).

**Review by jvilk-stripe — CHANGES_REQUESTED** [review]
Looks like the assertion still isn't testing the corner cases we discussed, but we're getting closer!

File: README.md [hunk 11]
```diff
@@ -652,6 +652,96 @@ class Project::Foo
...
+# typed: true
+# frozen_string_literal: true
+
+class Foo
+  class Foo
+  end
+end
+
+foo = Foo.new
+#     ^ apply-rename: [A] newName: foo invalid:true
```
[the remark this answers is not in the document]
Comment (vilhena-stripe) [reply README.md:708]: It is an invalid rename because the new name starts with a lower case letter

[the remark this answers is not in the document]
Comment (jvilk-stripe) [reply README.md:708]: Ah!! whoops

**Review by jvilk-stripe — APPROVED** [review]
Let's fix up those tests and I think this is ready to merge.
